File: codes/BOJ/DFS, BFS/7576.java
// BOJ 7576번: 토마토
import java.io.*;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int m, n;
    static int[][] grid, dist;
    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {1, -1, 0, 0};
    static Queue<Node> q = new LinkedList<>();
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        grid = new int[n][m];
        dist = new int[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                grid[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) q.add(new Node(i, j));
            }
        }
        while (!q.isEmpty()) {
            Node cur = q.poll();
            for (int i = 0; i < 4; i++) {
                int nx = cur.getX() + dx[i];
                int ny = cur.getY() + dy[i];
                if (nx >= 0 && ny >= 0 && nx < n && ny < m && grid[nx][ny] == 0) {
                    grid[nx][ny] = 1;
                    dist[nx][ny] = dist[cur.getX()][cur.getY()] + 1;
                    q.add(new Node(nx, ny));
                }
            }
        }

        int max = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 0) {
                    System.out.println(-1);
                    return;
                }
                max = Math.max(dist[i][j], max);
            }
        }
        System.out.println(max);
    }

    static class Node {
        private int x;
        private int y;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public int getX() {
            return x;
        }

        public int getY() {
            return y;
        }
    }
}

File: codes/BOJ/DFS, BFS/7576.py
# BOJ 7576번: 토마토
from collections import deque

dx = [0, 0, 1, -1]
dy = [1, -1, 0 ,0]

m, n = map(int, input().split())
tomatoes = [list(map(int, input().split())) for _ in range(n)]
dist = [[0] * m for _ in range(n)]

queue = deque()
for i in range(n):
    for j in range(m):
        if tomatoes[i][j] == 1:
            queue.append((i, j))

while queue:
    x, y = queue.popleft()
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx < n and 0 <= ny < m and tomatoes[nx][ny] == 0:
            tomatoes[nx][ny] = 1
            dist[nx][ny] = dist[x][y] + 1
            queue.append((nx, ny))

result = 0
for i in range(n):
    for j in range(m):
        if tomatoes[i][j] == 0:
            print(-1)
            exit()
        result = max(result, dist[i][j])

print(result)